controllers/namespaces_controller.go

```diff
@@ -0,0 +1,108 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"time"
+)
+
+// NamespacesReconciler reconciles a ClusterSecret object when new namespaces are created
+type NamespacesReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+//+kubebuilder:rbac:groups=chistadata.io,resources=clustersecrets,verbs=get;list;watch;create;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *NamespacesReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+
+	namespace := &v1.Namespace{}
+	err := r.Get(ctx, req.NamespacedName, namespace)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Namespace not found. Ignoring since object must be deleted", "namespace", namespace.Name)
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get Namespace, requeuing the request", "namespace", namespace.Name)
+		return ctrl.Result{RequeueAfter: time.Minute}, err
+	}
+
+	isMarkedToBeDeleted := namespace.GetDeletionTimestamp() != nil
+	if isMarkedToBeDeleted {
+		return ctrl.Result{}, nil
+	}
+
+	clusterSecrets := &v12.ClusterSecretList{}
+	err = r.List(ctx, clusterSecrets)
+	if err != nil {
+		return ctrl.Result{RequeueAfter: time.Minute}, err
+	}
+
+	for _, c := range clusterSecrets.Items {
+		now := time.Now()
+		if c.Annotations["chistadata.io/new-namespace-created-at"] != "" {
+			clusterSecretLastUpdateFromNamespace, err := time.Parse(time.RFC3339, c.Annotations["chistadata.io/new-namespace-created-at"])
+			if err != nil {
+				return ctrl.Result{RequeueAfter: time.Minute}, err
+			}
+			if now.Sub(clusterSecretLastUpdateFromNamespace) < time.Minute {
+				log.V(2).Info("The ClusterSecret was already updated by another namespace change", "clusterSecret.Name", c.Name)
+				continue
+			}
+		}
+		c.Annotations["chistadata.io/new-namespace-created-at"] = now.Format(time.RFC3339)
+
+		err = r.Update(ctx, &c)
+		if err != nil {
+			log.Error(err, "Failed to update ClusterSecret", "clusterSecret.Name", c.Name)
+			return ctrl.Result{RequeueAfter: time.Minute}, err
+		}
+		log.Info("ClusterSecret updated successfully", "clusterSecret.Name", c.Name)
+		time.Sleep(2 * time.Second)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *NamespacesReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Namespace{}).
+		WithEventFilter(predicate.Funcs{
+			CreateFunc: func(event event.CreateEvent) bool {
```

Comment:
IMHO predicates filter events from the workerqueue, what exactly is the predicate needed for ?

Reply:
This is to prevent every Namespace from triggering an update on the ClusterSecret during the startup time (first execution of reconciler). It would happen as the ClusterSecret is a cluster scoped resource and we have a loop of ClusterSecrets for each reconciliation of this controller and whenever we start the controller, it trigger a create event for all the existing Namespaces and this is handled by the ClusterSecretOperator at the startup, that's why they are being suppressed here.